src/decoder/zlib.rs

```diff
@@ -179,10 +187,41 @@ impl ZlibStream {
     }
 
     fn transfer_finished_data(&mut self, image_data: &mut Vec<u8>) -> usize {
-        let safe = self.out_pos.saturating_sub(CHUNCK_BUFFER_SIZE);
-        // TODO: allocation limits.
-        image_data.extend(self.out_buffer.drain(..safe));
-        self.out_pos -= safe;
-        safe
+        let transferred = &self.out_buffer[self.read_pos..self.out_pos];
+        image_data.extend_from_slice(transferred);
+        self.read_pos = self.out_pos;
+        transferred.len()
+    }
+
+    fn compact_out_buffer_if_needed(&mut self) {
+        // [PNG spec](https://www.w3.org/TR/2003/REC-PNG-20031110/#10Compression) says that
+        // "deflate/inflate compression with a sliding window (which is an upper bound on the
+        // distances appearing in the deflate stream) of at most 32768 bytes".
+        //
+        // `fdeflate` requires that we keep this many most recently decompressed bytes in the
+        // `out_buffer` - this allows referring back to them when handling "length and distance
+        // codes" in the deflate stream).
+        const LOOKBACK_SIZE: usize = 32768;
+
+        // Compact `self.out_buffer` when "needed".  Doing this conditionally helps to put an upper
+        // bound on the amortized cost of copying the data within `self.out_buffer`.
+        //
+        // TODO: The factor of 4 is an ad-hoc heuristic.  Consider measuring and using a different
+        // factor.  (Early experiments seem to indicate that factor of 4 is faster than a factor of
+        // 2 and 4 * `LOOKBACK_SIZE` seems like an acceptable memory trade-off.  Higher factors
+        // result in higher memory usage, but the compaction cost is lower - factor of 4 means
+        // that 1 byte gets copied during compaction for 3 decompressed bytes.)
+        if self.out_pos > LOOKBACK_SIZE * 4 {
+            // Only preserve the `lookback_buffer` and "throw away" the earlier prefix.
+            let lookback_buffer = self.out_pos.saturating_sub(LOOKBACK_SIZE)..self.out_pos;
+            let preserved_len = lookback_buffer.len();
+            self.out_buffer.copy_within(lookback_buffer, 0);
+            self.read_pos = preserved_len;
+            self.out_pos = preserved_len;
+
+            // We can't reuse the rest of `out_buffer` because it hasn't been zeroed-out (as
+            // expected by `fdeflate`.  Because of this, we `truncate` the `out_buffer`.
+            self.out_buffer.truncate(preserved_len);
```

Comment:
I eliminated the requirement in `fdeflate` about zeroing the buffer, so this shouldn't be needed

Reply:
Done - removed these 3 lines.

This changes how much will be decompressed in subsequent calls to `fn decompress` though - this probably means that I should redo the performance measurements to double-check the performance impact is still the same.